src/ctap2/commands/mod.rs

```diff
@@ -103,136 +102,137 @@ pub(crate) enum PinUvAuthResult {
-    }
-
-    fn handle_auth_command_errors<D: FidoDevice>(
-        &self,
-        dev: &mut D,
-        error: HIDError,
-        pin_uv_auth_result: PinUvAuthResult,
-        skip_uv: &mut bool,
-        status: &Sender<crate::StatusUpdate>,
-    ) -> Option<AuthenticatorError> {
-        match repackage_pin_errors(dev, error) {
-            AuthenticatorError::HIDError(HIDError::Command(CommandError::StatusCode(
-                StatusCode::ChannelBusy,
-                _,
-            ))) => {
-                // Channel busy. Client SHOULD retry the request after a short delay.
-                thread::sleep(Duration::from_millis(100));
-                None
-            }
-            // This should only happen for CTAP2.0 tokens that use internal UV and failed
-            // (e.g. wrong fingerprint used), while doing MakeCredentials
-            AuthenticatorError::PinError(PinError::PinAuthInvalid)
-                if pin_uv_auth_result == PinUvAuthResult::UsingInternalUv
-                    && Self::command() == Command::MakeCredentials =>
-            {
-                send_status(status, StatusUpdate::PinAuthInvalid);
-                None
-            }
-            // This should only happen for CTAP2.0 tokens that use internal UV and failed
-            // (e.g. wrong fingerprint used), while doing GetAssertion
-            // Yes, this is different than for MakeCredential.
-            AuthenticatorError::HIDError(HIDError::Command(CommandError::StatusCode(
-                StatusCode::OperationDenied,
-                _,
-            ))) if pin_uv_auth_result == PinUvAuthResult::UsingInternalUv
-                && Self::command() == Command::GetAssertion =>
-            {
-                send_status(status, StatusUpdate::PinAuthInvalid);
-                None
-            }
-            AuthenticatorError::PinError(PinError::PinRequired)
-                if pin_uv_auth_result == PinUvAuthResult::UsingInternalUv =>
-            {
-                // This should only happen for CTAP2.0 tokens that use internal UV and failed
-                // repeatedly, so that we have to fall back to PINs
-                *skip_uv = true;
-                None
-            }
-            err => Some(err),
-        }
+            let pin_auth_token = dev
+                .get_pin_token(self.pin())
+                .map_err(|e| repackage_pin_errors(dev, e))?;
+
+            // CTAP 2.0 spec is a bit vague here, but CTAP 2.1 is very specific, that the request
+            // should either include pinAuth OR uv=true, but not both at the same time.
+            // Do not set user_verification, if pinAuth is provided
+            self.set_uv_option(None);
+            (PinUvAuthResult::SuccessGetPinToken, pin_auth_token)
+        };
+
+        let pin_auth_param = pin_auth_token
+            .derive(self.client_data_hash().as_ref())
+            .map_err(CommandError::Crypto)?;
+        self.set_pin_auth(Some(pin_auth_param));
```

Comment:
Maybe this function should be called `set_pin_uv_auth_param`?

Reply:
Yes, it should.